### PR TITLE
Add Authenticators API implementation.

### PR DESCRIPTION
**Pull request by dewniMW**
## Purpose
Resolves https://github.com/wso2/product-is/issues/11984

## Goals

- Introduce a new API to retrieve all authenticators defined in identity server.
- Introduce a meta API to retrieve all tags defined for authenticators.

## Related PRs

- https://github.com/wso2/carbon-identity-framework/pull/3570
- https://github.com/wso2/carbon-identity-framework/pull/3577

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -481,7 +487,7 @@
         <maven.buildnumber.plugin.version>1.4</maven.buildnumber.plugin.version>
         <org.apache.felix.annotations.version>1.2.4</org.apache.felix.annotations.version>
         <identity.governance.version>1.4.75</identity.governance.version>
-        <carbon.identity.framework.version>5.19.20</carbon.identity.framework.version>
+        <carbon.identity.framework.version>5.20.93</carbon.identity.framework.version>
```
Comment (dewniMW) [inline pom.xml:490]: need to uodate the carbon identity framework version here after merging https://github.com/wso2/carbon-identity-framework/pull/3577

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  - url: 'https://localhost:9443/t/{tenant-domain}/api/server/v1'
+    variables:
+      tenant-domain:
+        default: carbon.super
+security:
+  - OAuth2: []
+  - BasicAuth: []
+
+paths:
+  /authenticators:
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml:23]: Lets provide limit pagination support in the definition. We can implement them later. Currently we can throw "Not Implemented" Error.

Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml:23]: Added with cc03f33ddc1e99cbfa0c5359ecf6f978c663e707

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.common/pom.xml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.wso2.carbon.identity.server.api</groupId>
+        <artifactId>org.wso2.carbon.identity.api.server.authenticators</artifactId>
+        <version>1.0.229-SNAPSHOT</version>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.common/pom.xml:25]: Make sure to update the latest pom version before merging.

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/pom.xml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  ~ under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.wso2.carbon.identity.server.api</groupId>
+        <artifactId>org.wso2.carbon.identity.api.server.authenticators</artifactId>
+        <version>1.0.229-SNAPSHOT</version>
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/pom.xml:26]: Make sure to update the latest pom version before merging.

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml [hunk 5]
```diff
@@ -0,0 +1,221 @@
+openapi: 3.0.0
+info:
+  version: "v1"
+  title: 'WSO2 Identity Server - Authenticators API Definition'
+  description: 'This document specifies a **RESTful API** for **WSO2 Identity Server Authenticators**'
+  contact:
+    name: WSO2
+    url: 'http://wso2.com/products/identity-server/'
+    email: architecture@wso2.org
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml:9]: Shall we change this to `iam-dev@wso2.org` as the `architecture@wso2.org` is deprecated now?

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+                }
+            }
+
+            LocalAuthenticatorConfig[] localAuthenticatorConfigs = AuthenticatorsServiceHolder.getInstance()
+                    .getApplicationManagementService().getAllLocalAuthenticators(ContextLoader
+                            .getTenantDomainFromContext());
+
+            int localAuthenticatorsCount = localAuthenticatorConfigs.length;
+            RequestPathAuthenticatorConfig[] requestPathAuthenticatorConfigs = new RequestPathAuthenticatorConfig[0];
+            if (localAuthenticatorsCount < maximumItemPerPage) {
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:103]: Why we need this check to get the requestPathAuthenticatorConfigs? Shall we add a comment by explaining the logic?

Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java]: fixed with e0d117b57a31fb8167d880e7dfc6c3f603702ff7

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml (hunk 5 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml]: fixed with e0d117b57a31fb8167d880e7dfc6c3f603702ff7

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 7]
```diff
@@ -0,0 +1,864 @@
...
+            LocalAuthenticatorConfig[] localAuthenticatorConfigs = AuthenticatorsServiceHolder.getInstance()
+                    .getApplicationManagementService().getAllLocalAuthenticators(ContextLoader
+                            .getTenantDomainFromContext());
+
+            int localAuthenticatorsCount = localAuthenticatorConfigs.length;
+            RequestPathAuthenticatorConfig[] requestPathAuthenticatorConfigs = new RequestPathAuthenticatorConfig[0];
+
+            // If there is no filter string available in the request, the request path authenticators are required to
+            // be fetched only if the  no. of local authenticators retrieved are less than the maximum items per page
+            // count as the no. of items returned in the response will be capped at the maximum items per page count.
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:106]: Shall we use multiline comments here?

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 8]
```diff
@@ -0,0 +1,864 @@
...
+            requestedAttributeList.add(Constants.FEDERATED_AUTHENTICATORS);
+
+            int idPCountToBeRetrieved = maximumItemPerPage - (localAuthenticatorsCount +
+                    requestPathAuthenticatorConfigs.length);
+            List<IdentityProvider> identityProviders = null;
+
+            // If there is no filter string available in the request, the identity providers are required to
+            // be fetched only if the total of local authenticators and request path authenticators retrieved above is
+            // less than the maximum items per page count as the no. of items returned in the response will be capped
+            // at the maximum items per page count.
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:123]: Shall we use multiline comments here?

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java (hunk 7 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java]: fixed

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java (hunk 8 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java]: fixed

**Comment by ashensw** [comment]
@dewniMW Since the framework is released let's bump the version to run the PR Builder.

**Review by ashensw — APPROVED** [review]

**Pull request merged**